### PR TITLE
Add `asyncio_tcp` benchmark

### PR DESCRIPTION
**Pull request by kumaraditya303**
From https://github.com/python/cpython/issues/91166

We can show off this optimization as part of 3.12 performance improvement :)

### CODE REVIEW

**Comment by brandtbucher** [comment]
Is the goal for this benchmark to guide our future improvements, or just to highlight the improvement already made in the linked issue?

We should really avoid the latter... the last thing I want is to add synthetic benchmarks just to boost our 3.12 numbers.

**Comment by kumaraditya303** [comment]
> Is the goal for this benchmark to guide our future improvements, or just to highlight the improvement already made in the linked issue?

It is to guide for future improvements and more importantly avoid regressions.  As it currently stands there are no benchmarks for `asyncio` thus it is hard to evaluate the impact of any bug fixes on performance, having this is `pyperformance` will help in tracking this.  (recent https://github.com/python/cpython/pull/100053)

> We should really avoid the latter... the last thing I want is to add synthetic benchmarks just to boost our 3.12 numbers.

That's true though this isn't any different from other benchmark, even if there is improvement in `pyperformance` it doesn't mean the user's application will be that much faster or faster at all. 

I am fine with keeping this in my own fork but I would prefer to have it in `pyperformance` so it can be tracked using speed.python.org.

**Comment by kumaraditya303** [comment]
FYI if this is acceptable then I have other benchmarks to add for `asyncio` notably `udp` and `ssl` performance.

**Comment by corona10** [comment]
Comment which does not judge this PR: I think that adding a new benchmark will consume the cost of running time for measuring.
Since we can not add every single benchmark for each dimension, each benchmark should cover a lot of dimensions of performance if possible.

**Comment by kumaraditya303** [comment]
> I think that adding a new benchmark will consume the cost of running time for measuring.

I was under the impression that cost of running the benchmark is not an issue. If it is an issue then it isn't worth it and I can run it myself when necessary so no need for this in pyperformance.

**Comment by ericsnowcurrently** [comment]
> I think that adding a new benchmark will consume the cost of running time for measuring.
> Since we can not add every single benchmark for each dimension, each benchmark should cover a lot of dimensions of performance if possible.

If running time is a concern, we can restrict the set of benchmarks that are run by default.  That is easily done in the [manifest file](https://github.com/python/pyperformance/blob/main/pyperformance/data-files/benchmarks/MANIFEST).  Then we can do longer runs whenever we feel comfortable with the extra time by passing the `--benchmarks all` (or maybe it is "<all>"?) option to pyperformance.

In the long term, ideally we would have a solid set of workload-oriented ("macro") benchmarks that measure real-world performance (close enough, at least).  This is what we would use to measure performance effectively and those results are what we would communicate to users.  However, we still need the wide variety of feature-oriented ("micro") benchmarks, to help us avoid regressions and narrow down causes when we have regressions, as well as point to where possible improvements might be.  See https://github.com/faster-cpython/ideas/wiki/All-About-Python-Benchmarking.

Currently, pretty much all of the pyperformance benchmarks are feature-oriented ("micro").  So we have plenty of work ahead of us.  (Note that the [pyston benchmarks](https://github.com/pyston/python-macrobenchmarks/tree/main/benchmarks) are much closer to workload-oriented and we have been incorporating them into the faster-cpython results as much as possible.)

**Comment by corona10** [comment]
> If running time is a concern, we can restrict the set of benchmarks that are run by default.

Ah okay, in that case, I am okay to add it.
I just worried about all various of bench_asyncio_xxxx.py.

**Comment by kumaraditya303** [comment]
Final benchmark result: 

| Benchmark   | 3.10 | patch                |
|-------------|:--------:|:--------------------:|
| asyncio_tcp | 1.41 sec | 724 ms: 1.95x faster |

I plan to merge this by weekend if there are no objections.

File: pyperformance/data-files/benchmarks/bm_asyncio_tcp/run_benchmark.py
Comment (gvanrossum) [inline pyperformance/data-files/benchmarks/bm_asyncio_tcp/run_benchmark.py:35]: I feel this should at least count the total number of bytes received and assert that it is as expected (100*CHUNK_SIZE?).

Comment (kumaraditya303) [reply pyperformance/data-files/benchmarks/bm_asyncio_tcp/run_benchmark.py:36]: done

Comment (kumaraditya303) [reply pyperformance/data-files/benchmarks/bm_asyncio_tcp/run_benchmark.py:36]: For some reason I didn't receive notification for your comment :/.

**Review by gvanrossum — APPROVED** [review]

**Comment by gvanrossum** [comment]
@mdboom What do you think?

**Comment by kumaraditya303** [comment]
I'll merge this now as https://github.com/python/cpython/pull/31871 is almost done, awaiting final review and merge. We can sort out any issues with this later down the road.

**Pull request merged**
